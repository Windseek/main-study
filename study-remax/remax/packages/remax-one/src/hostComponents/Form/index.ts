import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import defaults from './props/default';
import { FormProps } from './props';

export type { FormProps };

const Form: React.ComponentType<FormProps> = createHostComponent<FormProps>('form', null, defaults);

export default Form;
