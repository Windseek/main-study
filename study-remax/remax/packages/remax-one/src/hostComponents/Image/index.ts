import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import ImageProps from './props';
import defaults from './props/default';

export type { ImageProps };

const Image: React.ComponentType<ImageProps> = createHostComponent<ImageProps>('image', null, defaults);

export default Image;
