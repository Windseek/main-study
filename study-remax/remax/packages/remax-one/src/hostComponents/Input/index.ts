import * as React from 'react';
import memoizeOne from 'memoize-one';
import { InputEvent } from '../../types';
import { createCallback, createInputEvent, aliasProps } from '../../createHostComponent';
import type { InputProps } from './props';
import alias from './props/alias';

export type { InputProps };

export interface InputState {
  value?: string;
  controlled: boolean;
}

const createInputCallback = memoizeOne(createCallback);
const createConfirmCallback = memoizeOne(createCallback);
const createFocusCallback = memoizeOne(createCallback);
const createBlurCallback = memoizeOne(createCallback);

export default class Input extends React.Component<InputProps, InputState> {
  // 平台独有的属性默认值写在这
  static defaultProps = {
    'toutiao-selection-end': 999,
    'toutiao-selection-start': 999,

    'wechat-placeholder-class': 'input-placeholder',
    'wechat-cursor-spacing': 0,
    'wechat-confirm-type': 'done',
    'wechat-confirm-hold': false,
    'wechat-selection-end': -1,
    'wechat-selection-start': -1,
    'wechat-adjust-position': true,
    'wechat-hold-keyboard': false,
  };

  state: InputState = {
    value: '',
    // 阿里
    controlled: false,
  };

  constructor(props: InputProps) {
    super(props);

    const controlled = props.value !== undefined;
    const value = controlled ? props.value : props.defaultValue;

    this.state = {
      value,
      controlled,
    };
  }

  componentDidUpdate() {
    if (this.props.value !== undefined && this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  handleInput = (e: InputEvent) => {
    const { controlled } = this.state;

    if (typeof this.props.onInput === 'function') {
      return this.props.onInput(e);
    }

    // 微信
    // 注意，微信要对 input 受控，必须要在 onInput 方法返回值
    return controlled ? this.props.value : undefined;
  };

  render() {
    const inputProps = { ...this.props };

    if (inputProps.onInput) {
      inputProps.onInput = createInputCallback(this.handleInput, createInputEvent);
    }

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createConfirmCallback(this.props.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createFocusCallback(this.props.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createBlurCallback(this.props.onBlur, createInputEvent);
    }

    // 通用属性的默认属性根据平台在这里设置
    if (process.env.REMAX_PLATFORM === 'toutiao') {
      inputProps.maxLength = inputProps.maxLength ?? 140;
    }

    if (process.env.REMAX_PLATFORM === 'wechat') {
      inputProps.type = inputProps.type ?? 'text';
      inputProps.password = inputProps.password ?? false;
      inputProps.disabled = inputProps.disabled ?? false;
      inputProps.focus = inputProps.focus ?? false;
      inputProps.maxLength = inputProps.maxLength ?? 140;
    }

    const nextProps = aliasProps(
      {
        ...inputProps,
        ...this.state,
      },
      alias
    );

    return React.createElement('input', nextProps);
  }
}
