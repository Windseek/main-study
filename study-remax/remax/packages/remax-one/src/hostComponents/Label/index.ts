import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import { LabelProps } from './props';

export type { LabelProps };

const Label: React.ComponentType<LabelProps> = createHostComponent<LabelProps>('label', { htmlFor: 'for' });

export default Label;
