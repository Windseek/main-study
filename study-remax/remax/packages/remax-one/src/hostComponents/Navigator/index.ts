import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import { NavigatorProps } from './props';
import defaults from './props/default';

export type { NavigatorProps };

const Navigator: React.ComponentType<NavigatorProps> = createHostComponent<NavigatorProps>('navigator', null, defaults);

export default Navigator;
