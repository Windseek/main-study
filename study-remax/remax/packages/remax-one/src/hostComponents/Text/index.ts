import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import TextProps from './props';
import defaults from './props/default';

export type { TextProps };

const Text: React.ComponentType<TextProps> = createHostComponent<TextProps>('text', null, defaults);

export default Text;
