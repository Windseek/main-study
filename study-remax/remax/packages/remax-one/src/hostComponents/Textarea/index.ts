import * as React from 'react';
import { InputEvent } from '../../types';
import { createCallback, createInputEvent, aliasProps } from '../../createHostComponent';
import { TextareaProps } from './props';
import alias from './props/alias';

export type { TextareaProps };

export interface TextareaState {
  value?: string;
  controlled: boolean;
}

export default class Textarea extends React.Component<TextareaProps, TextareaState> {
  static defaultProps = {
    'toutiao-selection-end': 999,
    'toutiao-selection-start': 999,

    'wechat-selection-end': -1,
    'wechat-selection-start': -1,
    'wechat-fixed': false,
    'wechat-placeholder-class': 'textarea-placeholder',
    'wechat-cursor-spacing': 0,
    'wechat-cursor': -1,
    'wechat-show-confirm-bar': true,
    'wechat-adjust-position': true,
    'wechat-hold-keyboard': false,
    'wechat-disable-default-padding': false,
  };

  state: TextareaState = {
    value: '',
    // 阿里
    controlled: false,
  };

  constructor(props: TextareaProps) {
    super(props);

    const controlled = props.value !== undefined;
    const value = controlled ? props.value : props.defaultValue;

    this.state = {
      value,
      controlled,
    };
  }

  componentDidUpdate() {
    if (this.props.value !== undefined && this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  handleInput = (e: InputEvent) => {
    const { controlled } = this.state;

    if (!controlled) {
      this.setState({ value: e.target.value });
    }

    if (typeof this.props.onInput === 'function') {
      return this.props.onInput(e);
    }

    // 微信
    // 注意，微信要对 input 受控，必须要在 onInput 方法返回值
    return controlled ? this.props.value : e.target.value;
  };

  render() {
    const inputProps = {
      ...this.props,
      onInput: createCallback(this.handleInput, createInputEvent),
    };

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createCallback(this.props.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createCallback(this.props.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createCallback(this.props.onBlur, createInputEvent);
    }

    if (process.env.REMAX_PLATFORM === 'toutiao') {
      inputProps.maxLength = inputProps.maxLength ?? 140;
    }

    if (process.env.REMAX_PLATFORM === 'wechat') {
      inputProps.maxLength = inputProps.maxLength ?? 140;
      inputProps.disabled = inputProps.disabled ?? false;
      inputProps.focus = inputProps.focus ?? false;
      inputProps.autoHeight = inputProps.autoHeight ?? false;
    }

    const nextProps = aliasProps(
      {
        ...inputProps,
        ...this.state,
      },
      alias
    );

    return React.createElement('textarea', nextProps);
  }
}
