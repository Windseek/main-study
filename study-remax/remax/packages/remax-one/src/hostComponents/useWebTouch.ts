import * as React from 'react';
import { LONG_TAP_DURATION } from '../constants';

interface UseWebTouchParams {
  hoverDelay?: number;
  onLongTap?: (event: any) => void;
  onTouchCancel?: (event: any) => void;
  onTouchEnd?: (event: any) => void;
  onTouchMove?: (event: any) => void;
  onTouchStart?: (event: any) => void;
  hoverDuration?: number;
}

function useWebTouch<T>({
  hoverDuration,
  hoverDelay,
  onLongTap,
  onTouchCancel,
  onTouchEnd,
  onTouchMove,
  onTouchStart,
}: UseWebTouchParams): [
  boolean,
  (event: React.TouchEvent<T>) => void,
  (event: React.TouchEvent<T>) => void,
  (event: React.TouchEvent<T>) => void,
  (event: React.TouchEvent<T>) => void
] {
  const [touching, setTouching] = React.useState(false);
  const hoveringRef = React.useRef(false);
  const timers = React.useRef<number[]>([]);

  function executeTimeout(callback: () => void, time?: number) {
    const timer = setTimeout(() => {
      callback();
      timers.current.filter(t => t !== timer);
    }, time);
    timers.current.push(timer);
  }

  React.useEffect(
    () => () => {
      timers.current.forEach(t => clearTimeout(t));
    },
    []
  );

  function handleTouchStart(e: any) {
    hoveringRef.current = true;

    executeTimeout(() => {
      if (hoveringRef.current) {
        setTouching(true);
      }
    }, hoverDelay);

    executeTimeout(() => {
      if (hoveringRef.current && typeof onLongTap === 'function') {
        onLongTap(e);
      }
    }, LONG_TAP_DURATION);

    if (typeof onTouchStart === 'function') {
      return onTouchStart(e);
    }
  }

  function handleTouchMove(e: any) {
    hoveringRef.current = false;

    executeTimeout(() => {
      if (touching) {
        setTouching(false);
      }
    }, hoverDuration);

    if (typeof onTouchMove === 'function') {
      return onTouchMove(e);
    }
  }

  function handleTouchEnd(e: any) {
    hoveringRef.current = false;

    executeTimeout(() => {
      if (touching) {
        setTouching(false);
      }
    }, hoverDuration);

    if (typeof onTouchEnd === 'function') {
      return onTouchEnd(e);
    }
  }

  function handleTouchCancel(e: any) {
    hoveringRef.current = false;

    executeTimeout(() => {
      if (touching) {
        setTouching(false);
      }
    }, hoverDuration);

    if (typeof onTouchCancel === 'function') {
      return onTouchCancel(e);
    }
  }

  return [touching, handleTouchStart, handleTouchMove, handleTouchEnd, handleTouchCancel];
}

export default useWebTouch;
