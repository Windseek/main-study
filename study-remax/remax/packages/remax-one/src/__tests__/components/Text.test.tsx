import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import { Text } from '../../hostComponents';

describe('Text', () => {
  it('render correctly in ali', () => {
    process.env.REMAX_PLATFORM = 'ali';
    const testRenderer = TestRenderer.create(<Text className="class">text</Text>);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('render correctly in wechat', () => {
    process.env.REMAX_PLATFORM = 'wechat';
    const testRenderer = TestRenderer.create(<Text className="class">text</Text>);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('render correctly in toutiao', () => {
    process.env.REMAX_PLATFORM = 'toutiao';
    const testRenderer = TestRenderer.create(<Text className="class">text</Text>);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('render correctly in web', () => {
    process.env.REMAX_PLATFORM = 'web';
    const testRenderer = TestRenderer.create(<Text className="class">text</Text>);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
