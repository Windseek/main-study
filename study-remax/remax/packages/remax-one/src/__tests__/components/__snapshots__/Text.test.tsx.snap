// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text render correctly in ali 1`] = `
<text
  className="class"
>
  text
</text>
`;

exports[`Text render correctly in toutiao 1`] = `
<text
  className="class"
>
  text
</text>
`;

exports[`Text render correctly in web 1`] = `
<text
  className="class"
>
  text
</text>
`;

exports[`Text render correctly in wechat 1`] = `
<text
  className="class"
>
  text
</text>
`;
